### ISSUE TITLE
Couldn't connect to the collab server. Please reload the page and try again

### ISSUE THREAD

**Issue by shenlei190810** [excalidraw/excalidraw#4993]
I have deployed excalidraw project, run it on my server. Then I try to share with others, I found this info, I need your help, thank you.
![image](https://user-images.githubusercontent.com/3354569/160757835-279d22b9-1e3a-4696-ad87-78c03e01e7a1.png)

**Comment by yyj1983** [comment]
me too

**Comment by nxboo** [comment]
me too  too

**Comment by dwelle** [comment]
Hello. Are you running your own collab server (https://github.com/excalidraw/excalidraw-room)? If so, how are you setting up the env variables to point to your collab server?

**Comment by MathiasRenner** [comment]
Same here. I have the excalidraw-room/collab server at port 3002 on localhost up and running.
I tried to set the env vars, but still have the error shown in the first post here.

Could you please provide instructions how to set the env vars correctly?

**Comment by kucukkanat** [comment]
Hi, I had the same issue. 
Set `REACT_APP_WS_SERVER_URL` to `http://localhost:3002` solved it for me. 
@MathiasRenner

**Comment by yyj1983** [comment]
> 

Can you tell me which file you changed?

**Comment by MathiasRenner** [comment]
Same open question as @yyj1983 to @kucukkanat : 
I set this variable in the docker compose file as enviroment around [here](https://github.com/excalidraw/excalidraw/blob/master/docker-compose.yml#L17). Didn't work for me.
Any other hints?

**Comment by kucukkanat** [comment]
It's been a while but I think I changed this one : https://github.com/excalidraw/excalidraw/blob/master/.env.production

**Comment by Alfablos** [comment]
Hello,
first thing first this project is awesome!
Then, I have the same issue, setting the variable (tried master and v0.12) does not fix the issue. I've tried both using the prebuilt image from the docker hub and building it from the code as per README.

Here's my docker-compose.yml (one of them):
```yaml
version: "3.8"

services:
  excalidraw:
    image: excalidraw/excalidraw:latest
#    build:
#      context: .
#      args:
#        - NODE_ENV=development
    container_name: excalidraw
    ports:
      - "3000:80"
    restart: on-failure
    stdin_open: true
    healthcheck:
      disable: true
    environment:
      NODE_ENV: development
      # REACT_APP_WS_SERVER_URL: http://localhost:3002               # Doesn't work either.
      REACT_APP_WS_SERVER_URL: http://excalidraw-room:3002

  excalidraw-room:
    image: excalidraw/excalidraw-room
    environment:
      # Collaboration WebSocket
      PORT: 3002
    ports:
      - "5000:80"
      - "3002:3002"

```

I checked the logs from the "room" service, nothing gets logged.
However, the logs from the excalidraw service show something interesting:
```
excalidraw  | 2022/10/22 12:44:17 [error] 32#32: *10 open() "/usr/share/nginx/html/undefined/collab-server" failed (2: No such file or directory), client: 172.21.0.1, server: localhost, request: "GET /undefined/collab-server HTTP/1.1", host: "localhost:3000", referrer: "http://localhost:3000/"
```

As a matter of fact the file does not exist, also I'm not sure why nginx is trying to access it so I'm just reporting this.


I hope this helps troubleshooting this until the collaboration server gets integrated (which would be awesome!).

Thanks guys!

**Comment by PappyEric** [comment]
Just installed and having this same issue. Was there a solution posted somewhere? Seems odd that a whiteboard collaboration server can't do collaboration.

**Comment by tiran133** [comment]
In case someone needs this still.

You need to build your own docker image with either `ENV VITE_APP_WS_SERVER_URL=http://127.0.0.1:3002` the Dockerfile

OR

change it in `.env.production`


This is how i did it.

```shell 
mkdir excalidraw
cd excalidraw
git clone git clone https://github.com/excalidraw/excalidraw-room.git
git clone https://github.com/excalidraw/excalidraw.git
rm excalidraw/.dockerignore
touch docker-compose.yaml
```

docker-compose.yaml

```yaml
version: "3.8"

services:
  excalidraw:
    build:
      context: ./excalidraw
      args:
        - NODE_ENV=production
    container_name: excalidraw
    ports:
      - "3000:80"
    restart: on-failure
    stdin_open: true
    healthcheck:
      disable: true
    
  excalidraw-room:
    build:
      context: ./excalidraw-room
    environment:
      # Collaboration WebSocket
      PORT: 3002
    ports:
      - "3002:3002"
```

Go into `excalidraw/.env.production` and change `VITE_APP_WS_SERVER_URL=` to  
`VITE_APP_WS_SERVER_URL=http://127.0.0.1:3002`

Make sure that `VITE_APP_WS_SERVER_URL` url can be reached from your host. The browser is connecting to it.

```shell
docker compose up -d --build
```

Browse to http://127.0.0.1:3000 and start collaborating.

**Comment by Pradur241** [comment]
Hi, tiran133. Thanks for your solution. It helps a lot. After deploying according to the way you gave me, I found that I can only access the collaboration feature on the local machine. How do I modify the configuration file section if I want to share it on a local area network in my office? @tiran133

**Comment by Pradur241** [comment]
Thanks! I will make a try.

**Comment by tiran133** [comment]
No this should be fine as long as you only change the first part of the port in the composer file

```yaml
 ports:
      - "10000:80"
```
If you want to change 3002 also you need to adjust the `VITE_APP_WS_SERVER_URL` accordingly.

**Comment by Pradur241** [comment]
Thanks a lot!! @tiran133

**Comment by flintstones-fred** [comment]
> In case someone needs this still.
> 
> You need to build your own docker image with either `ENV VITE_APP_WS_SERVER_URL=http://127.0.0.1:3002` the Dockerfile
> 
> OR
> 
> change it in `.env.production`
> 
> This is how i did it.
> 
> ```shell
> mkdir excalidraw
> cd excalidraw
> git clone git clone https://github.com/excalidraw/excalidraw-room.git
> git clone https://github.com/excalidraw/excalidraw.git
> rm excalidraw/.dockerignore
> touch docker-compose.yaml
> ```
> 
> docker-compose.yaml
> 
> ```yaml
> version: "3.8"
> 
> services:
>   excalidraw:
>     build:
>       context: ./excalidraw
>       args:
>         - NODE_ENV=production
>     container_name: excalidraw
>     ports:
>       - "3000:80"
>     restart: on-failure
>     stdin_open: true
>     healthcheck:
>       disable: true
>     
>   excalidraw-room:
>     build:
>       context: ./excalidraw-room
>     environment:
>       # Collaboration WebSocket
>       PORT: 3002
>     ports:
>       - "3002:3002"
> ```
> 
> Go into `excalidraw/.env.production` and change `VITE_APP_WS_SERVER_URL=` to `VITE_APP_WS_SERVER_URL=http://127.0.0.1:3002`
> 
> Make sure that `VITE_APP_WS_SERVER_URL` url can be reached from your host. The browser is connecting to it.
> 
> ```shell
> docker compose up -d --build
> ```
> 
> Browse to http://127.0.0.1:3000 and start collaborating.

Thank you @tiran133 for documenting the solution. I am trying to implement this and struggling with locating `excalidraw/.env.production`. Is this a file that I need to create?

Also, just FYI (since many others may end up following your guidelines) - in the `git clone` command, there is an extra `git clone`

Thanks!

**Comment by flintstones-fred** [comment]
@tiran133 Thank you for providing the instructions. When I follow these, however, the build fails. Can you see if I am missing something? Here is the output I get - 

```
docker-compose up -d --build
Building excalidraw
Sending build context to Docker daemon  7.091MB
Step 1/10 : FROM node:18 AS build
 ---> ef847f8b5ae3
Step 2/10 : WORKDIR /opt/node_app
 ---> Using cache
 ---> a8643ec15466
Step 3/10 : COPY package.json yarn.lock ./
 ---> Using cache
 ---> 3332f5298f0a
Step 4/10 : RUN yarn --ignore-optional --network-timeout 600000
 ---> Using cache
 ---> 8253551f785d
Step 5/10 : ARG NODE_ENV=production
 ---> Using cache
 ---> 094d94fdd98f
Step 6/10 : COPY . .
 ---> Using cache
 ---> b619ce293339
Step 7/10 : RUN yarn build:app:docker
 ---> Running in 57aa6628b835
yarn run v1.22.19
$ cross-env VITE_APP_DISABLE_SENTRY=true VITE_APP_DISABLE_TRACKING=true vite build
/bin/sh: 1: cross-env: not found
error Command failed with exit code 127.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
The command '/bin/sh -c yarn build:app:docker' returned a non-zero code: 127
ERROR: Service 'excalidraw' failed to build : Build failed
```

**Comment by mckennajones** [comment]
@eakteam I found this PR from @JannikStreek that addresses the issue: https://github.com/excalidraw/excalidraw/pull/7502

I copied the changes from that branch into my excalidraw Dockerfile and it builds successfully. Thanks @JannikStreek!

**Comment by DineshCodeFlow** [comment]
> For everyone who is having troubles here he an awesome repository which works like a charm :)
> 
> https://github.com/Nenodema/excalidraw-self-hosted-stack

Broken link!

**Comment by d3xt3r01** [comment]
So there's no way other than rebuilding the whole container?
I'd imagine the current image would just read the ENV on start and use that. Why rebuild everything for an url env change?

**Comment by wiktormalyska** [comment]
> Since the docker compose file exposes the ports to the host via the ports section. So there are available on localhost and on the host IP
> 
>     ports:
>       - "3000:80"
>     
>      ports:
>       - "3002:80"
> Instead of `VITE_APP_WS_SERVER_URL=http://127.0.0.1:3002` just replace the 127.0.0.1 with the hosts IP. for ex. `VITE_APP_WS_SERVER_URL=http://192.168.1.42:3002`
> 
> Now you can connect to http://192.168.1.42:3000. It should work from any computer within the same network.

Hi what if i use docker network?
```
services:
  excalidraw:
    container_name: excalidraw
    build:
      context: ./excalidraw
      args:
        - NODE_ENV=production
    restart: on-failure
    stdin_open: true
    healthcheck:
      disable: true
    networks:
      - xcaddy
      - excalidraw
  excalidraw-room:
    container_name: excalidraw-room
    build:
      context: ./excalidraw-room
    environment:
      PORT: 3002
    networks:
      - excalidraw
networks:
  excalidraw: null
  xcaddy:
    external: true
```

I also use xcaddy for proxy and i've set .env.production `VITE_APP_WS_SERVER_URL=http://excalidraw-room:3002`

**Comment by madukan** [comment]
Fabulous [instructions](https://github.com/excalidraw/excalidraw/issues/4993#issuecomment-1783669768) there @tiran133! Thanks and much appreciated!!

I was able to get localhost:3000 working right out of the repositories with your Docker compose configuration and VITE_APP_WS_SERVER_URL update. I can see the avatars moving, shapes creation etc by using two browser sessions! 

Now I have a tiny issue (which I assume others have already resolved?) - when I use a host name or an IP the collaboration feature doesn't work. I tested both with a host file entry and a local network IP.

What happens is when I used either, collaboration feature doesn't work. 

Going to the `http://<ip/host>:3002` URL shows the message (I ensure the configuration in .env.production matches what I access on browser):
```
Excalidraw collaboration server is up :)
```

Browser console has this exception and that gets logged when I click the "Start Session" button.
```
encryption.ts:17 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'generateKey')
    at CT (encryption.ts:17:42)
    at Rxe (index.ts:151:25)
    at async Object.startCollaboration (Collab.tsx:1009:7)
```

Any ideas/tips what may be going wrong here?

**Comment by madukan** [comment]
Found the answer to my own [question above](https://github.com/excalidraw/excalidraw/issues/4993#issuecomment-3704886633). 

Apparently it is a browser security issue (not a real issue) related. If I disable browser security 

My Chrome browser security was in "Standard Protection". Pushing that to "No Protection" helped eliminate this issue and test without proper SSL certificates (I was using self-signed).

Here's the Chrome settings screenshot (NOTE: THIS IS **ABSOLUTELY NOT** RECOMMENDED, I used just for testing my own deployment locally only):
<img width="1046" height="678" alt="Image" src="https://github.com/user-attachments/assets/18e596c7-d456-4053-9f15-c605228c6a69" />

Best way to solve this issue is to generate SSL certificates (or use a tool like `mkcert` to add a CA (Certificate Authority) to your OS (in MacOS, Mkcert adds the CA to Keychain automatically), while generating certificates for your Nginx proxy, or similar).

So to sum up, this issue is nothing to do with Excalidraw nor NodeJS. Using appropriate SSL certificates to serve resources that Excalidraw use to encrypt the payload solves the issue.

**Comment by l4rm4nd** [comment]
> So there's no way other than rebuilding the whole container?
> I'd imagine the current image would just read the ENV on start and use that. Why rebuild everything for an url env change?

There is another solution.

We can adjust the entrypoint and fix the hardcoded room URLs for collaboration.

See https://github.com/excalidraw/excalidraw/issues/5188#issuecomment-3848252703
